README
WARNING: this is code from 2002. In cleanup process.


                             Welcome to MODEL,
             the Modest Ordinary Differential Equation Library
                                      
                              Michael Peeters
   
                                   Overview
                                       
   MODEL is a numeric simulation library written during my PhD to
   simulate systems of rate equations describing Vertical Cavity Surface
   Emitting Lasers (VCSELs). I decided to write one myself after looking
   at existing libraries and deciding that they were either too
   complicated to use (having a target user base of mathematicians) or
   too opaque (i like to know what the code is doing, exactly).
   
   Since most of my programming before has been done in C/C++, what you
   see here is a C++ library composed of various interacting classes
   which have the following main functionalities:
   
     * deterministic integration of any system of well-behaved
       differential equations.
     * stochastic integration of these systems, with the possibility of
       specifying the correlations present in the noise
     * nonlinear rootfinder, to find stationary solutions
     * eigenvalue determination, for stability analysis
     * easy time modulation of input parameters
     * diverse data collecting classes for data analysis
     * Small signal analysis (first order)
       
   The following will be added RSN:
   
     * Fourier transforms, for spectral analysis (although this can be
       done in an external program)
     * Periodic solution finder
       
   It furthermore provides a numerical vector class, vectorfunction
   classes, LU solver and random generators.
   
   As I developed it using publicly available resources, GNU/Linux and
   other GPL'ed software, I decided that it should be GPL as well.
   However (Oh no, a "however" ! Let's hope it does not invalidate the
   copyleft), I would very much appreciate it if you let me know if you
   have used MODEL in any of your applications/simulations/research and
   provide a reference (this way, I can refer to your work, too).
   
   At the moment, MODEL has the rather arbitrary version number 1.0.
   Meaning it is useful. Period. Some interfaces (especially the
   stochastics) might still change, and I would like to add some ieee
   floating point exception trapping to avoid silly numerical errors.
   
                                 Installation
                                       
   Assuming you downloaded a full .tar of checked out the full CVS tree,
   you first have to make sure everything is configured correctly on your
   system: run ./configure to generate the correct configuration. Then a
   simple make or gnumake (on some systems) will do the trick. If you
   wish to change the option submitted to the compiler, define the
   environment variable CXXFLAGS to contain those you need (I know there
   must be a better way to do this). To get it to compile on the Alpha
   cluster here (www.vub.ac.be/bfucc , I have to export
   CXXFLAGS="-mieee-malpha-as" before running the configuration script.
   
   Once the make process has ended, You should have a libModel.a in the
   model directory and a singlemode executable in /tutorial. Run it to
   see if all went well (make test should do the trick). Learn to use
   gnuplot :-): plot "stepmodulation.dat" u 1:3 w l and admire the
   relaxation oscillations.
   
                                 Documentation
                                       
   If you want to generate the introduction, API and tutorial
   documentation, use make docs. This assumes you have a full teTex
   distribution, Doxygen and a copy of lgrind to prettyprint the code,
   however. Your mileage may vary. You can also regenerate the README
   file by typing make README. YMMVAL. You can generate an introductory
   PDF file using make pdf.
   


model/ChangeLog
2002-11-21 10:47  mpeeters

	* Makefile.in, cowner.h, eigenvalues.h, integrator.h, invariant.h,
	lusolve.h, newtonroot.h, normfunction.h, odesystem.h, rootscan.h,
	ssa.h, timeframe.h, vfwithbump.h (utags: gcc3_2-20021121): Modified
	C++ code to compile under gcc 3.2 (implicit typenames removed,
	slist in different namespace).

2002-11-19 23:18  mpeeters

	* Makefile.am (tags: gcc3_2-20021121), Makefile.in, sinmod.cpp
	(tags: gcc3_2-20021121), sinmod.h (tags: gcc3_2-20021121): Added
	sinmod, a sinus wave modulator (thanks guy).

2002-11-18 15:38  mpeeters

	* Makefile.am, Makefile.in (utags: STABLE): Forgot to remove
	ssa.cpp dependency in Makefile! Whoops.

2002-11-18 15:29  mpeeters

	* ChangeLog (tags: gcc3_2-20021121, STABLE): Updated docs and
	ChangeLog

2002-11-18 15:10  mpeeters

	* rootscan.h (tags: STABLE), ssa.cpp, ssa.h (tags: STABLE): Removed
	unnecessary ssa.cpp. Added DEBUG in rootscan catch{} routine. Added
	RC and phase info to the SSA routines.

2002-09-09 15:02  mpeeters

	* Makefile.am, Makefile.in, ssa.cpp, ssa.h
	(feature-blitz_tinyvector-2002-08): Removed ssa.cpp (again).
	Correct norm() error in ssa.h - but the wrong way. Drats...

2002-09-09 14:59  mpeeters

	* ssa.cpp (feature-blitz_tinyvector-2002-08): Somehow ssa.cpp was
	removed. Drats - of course: it was empty - but we should prune it
	from the makefile !!! So - next comment will be: it is removed.

2002-09-06 00:10  mpeeters

	* numvector.h, ssa.h (feature-blitz_tinyvector-2002-08): Changed
	norm to Mnorm. Added note.

2002-09-06 00:08  mpeeters

	* ssa.cpp (feature-blitz_tinyvector-2002-08): File not used. 
	Changed norm to Mnorm.

2002-09-04 10:17  mpeeters

	* Makefile.in, integrator.h (feature-blitz_tinyvector-2002-08):
	Added optimal optimalisation settings, rewrite of major integrators
	to take into account the new features of the TinyVector

2002-09-04 00:19  mpeeters

	* Makefile.in, newtonroot.h, normfunction.h, numvector.h,
	numvectorprint.h, rootscan.h (feature-blitz_tinyvector-2002-08): It
	works. Woohoo. Now see if it is faster.

2002-09-03 16:27  mpeeters

	* Makefile.am, Makefile.in, invariant.h, linesearch.h, lusolve.h,
	newtonroot.h, numerictraits.h, numerictypes.h, numvector.h,
	numvectorprint.h, rootscan.h (feature-blitz_tinyvector-2002-08):
	Changed to blitz TinyVector<>. "make test" compiles, but segfault
	just before the small signal analysis.

2002-07-31 13:28  mpeeters

	* bin.cpp, bin.h, Makefile.am, Makefile.in, binprobe.h,
	dwellprobe.h, integrator.h, lusolve.h, newtonroot.h,
	numvectorprint.h, odesystem.h, probe.h, random.cpp, random.h,
	rootscan.h, ticktock.h, vectorfunction.h, ChangeLog
	(feature-parameterlist-2002-08): Merging the old orphaned branch.

2002-07-31 12:51  mpeeters

	* Makefile.am, Makefile.in, bin.cpp, bin.h, binprobe.h,
	dwellprobe.h, integrator.h, lusolve.h, modulator.h, newtonroot.h,
	numvectorprint.h, odesystem.h, parameters.h, parmlist.h, probe.h,
	random.cpp, random.h, rootscan.h, ssa.h, ticktock.h,
	vectorfunction.h (feature-parameterlist-2002-08) (utags:
	admin-insert_ancient_history-2002_07_31-post): Committing changes
	from an olf orphaned branch.

2002-07-30 21:57  mpeeters

	* ChangeLog (tags: admin-insert_ancient_history-2002_07_31-post),
	Makefile.am, Makefile.in, bin.cpp (tags: gcc3_2-20021121, STABLE),
	bin.h (tags: gcc3_2-20021121, STABLE), bin2D.cpp (tags:
	gcc3_2-20021121, admin-insert_ancient_history-2002_07_31-post,
	STABLE), bin2D.h (tags: gcc3_2-20021121,
	admin-insert_ancient_history-2002_07_31-post, STABLE), binprobe.h
	(tags: gcc3_2-20021121, STABLE), dwellprobe.h (tags:
	gcc3_2-20021121, STABLE), integrator.h (tags: STABLE), lusolve.h
	(tags: STABLE), newtonroot.h (tags: STABLE), numvectorprint.h
	(tags: gcc3_2-20021121, STABLE), odesystem.h (tags: STABLE),
	probe.h (tags: gcc3_2-20021121, STABLE), random.cpp (tags:
	gcc3_2-20021121, STABLE), random.h (tags: gcc3_2-20021121, STABLE),
	rootscan.h, ssa.h, ticktock.h (tags: gcc3_2-20021121, STABLE),
	vectorfunction.h (tags: gcc3_2-20021121, STABLE) (utags:
	release-1_0_3-2002_08, trunkmerge-2002_08,
	admin-insert_ancient_history-2002_07_31-pre): Merge of 1.0.2 branch
	into trunk. From now on, the trunk will be a stable (meaning it
	compiles and has no quirks) branch and all special stuff will be
	done on separate branches waiting to be merged into the stable one.
	A tag STABLE will be made which moves with the trunk. Also, the
	main trunk will be considered to be 1.0.3.

2002-07-30 10:59  mpeeters

	* ChangeLog, vectorfunction.h (release-1_0_2-2001_08-stable)
	(utags: branchmerge-2002_08): Spring cleaning commit -
	administrative stuff.

2002-07-30 10:39  mpeeters

	* bin.cpp, bin.h, dwellprobe.h (release-1_0_2-2001_08-stable)
	(utags: branchmerge-2002_08): Added pdf computation to binning
	stuff (bob).

2002-07-24 12:01  mpeeters

	* random.cpp, random.h (release-1_0_2-2001_08-stable) (utags:
	branchmerge-2002_08): Changed routines to prevent duplication of
	seed (but not in an efficient way).

2002-07-24 11:59  mpeeters

	* odesystem.h (tags: branchmerge-2002_08)
	(release-1_0_2-2001_08-stable): Added possibility to start without
	predefined starting point.

2002-07-24 11:58  mpeeters

	* dwellprobe.h (release-1_0_2-2001_08-stable): New AVGAVGDwellprobe
	that computes the average dwelltime of an averaged out trace.

2002-07-24 11:56  mpeeters

	* Makefile.in (tags: branchmerge-2002_08)
	(release-1_0_2-2001_08-stable): Added bin2D.* dependency.

2002-07-24 11:55  mpeeters

	* binprobe.h (tags: branchmerge-2002_08)
	(release-1_0_2-2001_08-stable): Better output format.

2002-07-22 23:24  mpeeters

	* ticktock.h (tags: branchmerge-2002_08)
	(release-1_0_2-2001_08-stable): Defined a default implementation
	for tick(), so the TickTock class can be used as a way to define a
	timeframe hierarchy.

2002-07-09 10:40  mpeeters

	* newtonroot.h (tags: branchmerge-2002_08)
	(release-1_0_2-2001_08-stable): Added explicit type conversion for

model/ssa.h
/***************************************************************************
			ssa.h
			-----------
                             
    begin                : Tue Apr 24 17:56:51 CEST 2001
    author               : (C) 2001 by Michael Peeters
    email                : Michael.Peeters@vub.ac.be
***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifndef SSA_H
#define SSA_H

#include "utility.h"
#include <stdexcept>
#include "numerictraits.h"
#include "numerictypes.h"
#include "jacobian.h"
#include "rootscan.h"

namespace MODEL {

  /** Feed me a Vectorfunction and I can give you a small signal
	  analysis for any parameter. Actually, I will calculate it for
	  all parameters */

  template <integer dims, typename nelem=number, class NT = NumericTraits<nelem,dims> >
  class SSA
  {
  public:
	typedef typename NT::number	numT;
	typedef	typename NT::vect	vect;	
	typedef typename NT::matrix	matrix;
	typedef typename Jacobian<2*dims>::matrix system;
	typedef typename Jacobian<2*dims>::vect input;	
	typedef typename NumericTraits<complex,dims>::vect response;
	typedef response cinput;
	typedef typename NT::vf						vf;
	
  public:
	/** For my_sys, for this input parameter */
	SSA(vf& my_sys, const string& parm);

	/** Set the stationary parameter value around which we will
		modulate. This will return the number of stable stationary points the
		system have been able to find. Use get_stat_points to get a
		list in ScanList format */
	counter set_stat_param(const number& value);
	
	/** Once you have some points, choose one */
	void set_stat_point(const counter& po=1);

	/** Or (DANGEROUS) set it explicitly. Nobody guarantees this will
		work (and give meaningfull results) if you do not have a valid point */
	void set_stat_point(const vect& right_there);

	/** get a list of the stationary points */
	const ScanList<dims,NT>& get_stat_points(void);
	
	/** get a single response point. For that detailed analysis */
	response calc_point(const number& omega);
	
	/** get a list of responses, equally spaced in a log scale */
	ScanList<dims,complex> calc_response(const number& from, const
										 number& to, const counter&
										 n);

	/** get a list of responses real,complex, equally spaced in a log scale */
	ScanList<dims> calc_responseRC(const number& from, const
										 number& to, const counter&
										 n);

	/** get a list of responses, but this this just the norm, no phase
		information, please */
	ScanList<dims> calc_response_norm(const number& from, const
									  number& to, const counter&
									  n);

	/** get a list of responses, but this this just the phase, no amplitude
		information, please */
	ScanList<dims> calc_response_phase(const number& from, const
									  number& to, const counter&
									  n);

	/** set the depency manually. This is needed if we want to use
		these routines for other purposes, like noise analysis, where
		the sources are not necessarily the inputs */
	void set_dep(const vect& input)
	{
	  linpar = 0.;
	  for (integer i=0;i<dims;i++)
		linpar[i]=input[i];
	}

	/** set the depency manually when it is complex. This is needed if we want to use
		these routines for other purposes, like noise analysis, where
		the sources are not necessarily the inputs */
	void set_dep(const cinput& input)
	{
	  linpar = 0.;
	  for (integer i=0;i<dims;i++)
		{
		  linpar[i]=input[i].real();
		  linpar[dims+i]=input[i].imag();
		}	  
	}
  
  private:
	/** Calculate the parameter dependency: partial FD */
	void calc_dep(void);

	/** fill omega depency in system matrix A */
	void put_omega(const number& omega);

  private:
	/** Tells all calculations that all is well and they can go ahead
	 */
	bool all_is_well;

	/** The function/jacobian in question */
	Jacobian<dims>* J; 

	/** The actual jacobian of the vectorfunction */
	matrix j; // small j: nice and confusing
	
	/** A matrix twice as large for the complex problem */
	system A;
	
	/** The rootscanner */
	RootScan<dims,nelem>* rs;

	/** The list of stationary points */
	ScanList<dims,nelem> statp;

	/** The stationary point */
	vect here;

	/** The parameter */
	ParameterP p;

	/** The linear approximation of the way the equations respond to
		the input. Also twice as large, but only the top half in used  */
	input linpar;

  private:
	// to avoid silly errors
	NO_COPY(SSA);
	
  };
  
  // ------------------------------------------------------------

  template <integer dims, typename nelem, class NT >
  SSA<dims,nelem,NT>::SSA(vf& my_sys, const string& parm)
  {
	all_is_well=false;
	
	// get the jacobian, copy the function as is
	J=new Jacobian<dims>(my_sys,1E-6,true);

	// get the parameter
	p=&(J->get_function().get_parameter(parm));

	// init the scanner
	rs=new RootScan<dims,nelem>(J->get_function());	
  }

  template <integer dims, typename nelem, class NT >
  counter SSA<dims,nelem,NT>::set_stat_param(const number& value)
  {
	*p=value;

	statp=rs->scan(0.,1.,1); // these are dummy values
	statp=statp.select(RootScan<dims,nelem>::stable);

	// statp.print_raw(cout);

	// cerr << "SSA: started number of solutions " << statp.get_data()[0.].size() << endl;

	return statp.get_data()[0.].size();
	// watch out here: you might find only one (wrong point), because
	// we are NOT scanning. This is a problem in RootScan.
  }
  
  template <integer dims, typename nelem, class NT >
  void SSA<dims,nelem,NT>::set_stat_point(const counter& po)
  {
	// go to correct point
	if(statp.get_data()[0.].size()>0)
	  {
		// take the first parameter value in the list
		typename ScanList<dims,nelem>::parpointlist::iterator
		  getit=statp.get_data()[0].begin();
	
		// go to the right value
		for(counter i=po;i>1;i--) getit++;
		
		// the first value is the point...
		here=(*getit)[0];
		
		// now that we know where we, calculate the approximations
		calc_dep();

		// now we can do something
		all_is_well=true;		
	  }
	
  }
  template <integer dims, typename nelem, class NT >
  void SSA<dims,nelem,NT>::set_stat_point(const vect& right_there)
  {
		here=right_there;
		
		// now that we know where we, calculate the approximations
		calc_dep();

		// now we can do something
		all_is_well=true;
  }
  
  template <integer dims, typename nelem, class NT >
  typename SSA<dims,nelem,NT>::response SSA<dims,nelem, NT>::calc_point(const number& omega)
  {
	if(all_is_well) {
	  
	  put_omega(omega);
	  LUSolve<2*dims> solA(A,true); // make a copy
	
	  // copy into correct form
	  input t1 (solA(linpar));
	  response t2;
	  for (integer i=0;i<dims;i++)
		{
		  t2[i]=t1[i]+I*t1[i+dims];
		}

	  return t2;
	}
	
  }

  template <integer dims, typename nelem, class NT >
  ScanList<dims,complex>  
  SSA<dims,nelem, NT>::calc_response(const number& from, const
									 number& to, const counter&
									 n)
  {
	if(all_is_well) {
	  
	  ScanList<dims,complex> res;

	  number scaler=pow(to/from,1./number(n));
	  for (number omega=from;omega<to;omega*=scaler)
		{
		  res.add_point(calc_point(omega));
		  res.add_param(omega);
		}
	  return res;
	}
	
  }

  template <integer dims, typename nelem, class NT >
  ScanList<dims>
  SSA<dims,nelem, NT>::calc_responseRC(const number& from, const
									 number& to, const counter&
									 n)
  {
	if(all_is_well) {

	  ScanList<dims> res;

	  number scaler=pow(to/from,1./number(n));
	  for (number omega=from;omega<to;omega*=scaler)
		{
			response full;
			full=calc_point(omega);

			vect re,im;

			for(int i=0;i<dims;i++)
			{
				re[i]=real(full[i]);
				im[i]=imag(full[i]);
			}

		  res.add_point(re);
		  res.add_point(im);
		  res.add_param(omega);
		}
	  return res;
	}

  }

template <integer dims, typename nelem, class NT >
  ScanList<dims>
  SSA<dims,nelem, NT>::calc_response_norm(const number& from, const
										  number& to, const counter&
										  n)
  {
	if(all_is_well) {
	  
	  ScanList<dims> res;

	  number scaler=pow(to/from,1./number(n));
	  for (number omega=from;omega<to;omega*=scaler)
		{
		  response t1=calc_point(omega);
		  vect t2;
		  for(integer i=0;i<dims;i++)
			t2[i]=norm(t1[i]);

		  res.add_point(t2);
		  res.add_param(omega);
		}
	  return res;
	}

  }

  template <integer dims, typename nelem, class NT >
  ScanList<dims>  
  SSA<dims,nelem, NT>::calc_response_phase(const number& from, const
										  number& to, const counter&
										  n)
  {
	if(all_is_well) {
	  
	  ScanList<dims> res;

	  number scaler=pow(to/from,1./number(n));
	  for (number omega=from;omega<to;omega*=scaler)
		{
		  response t1=calc_point(omega);
		  vect t2;
		  for(integer i=0;i<dims;i++)
			t2[i]=arg(t1[i]);

		  res.add_point(t2);
		  res.add_param(omega);
		}
	  return res;
	}
	
  }
  
  template <integer dims, typename nelem, class NT >
  void SSA<dims,nelem, NT>::calc_dep(void)
  {
	// generate the Jacobian accurately (slow)
	j=J->calculate_accurate(here);

	// And copy it to the larger system
	for (int i=0;i<dims;i++)
	  {
		for (int k=0;k<dims;k++)
		  {
			A[i][k] = -j[i][k];
			A[dims+i][dims+k] =-j[i][k];
		  }  
	  }
	
	// find the (linear) dependecies of the equations to the
	// (tiny/small/microscopic input). This is done in a way very
	// similar to the Jacobian.
	// it is store in linpar
	linpar=0.; // just to be safe

	// value now
	vf& f=J->get_function();
	vect    fu ( f(here) );

	number pdp=*p;

	/** \todo this should be modifiable */
	const number epsilon = 1;
	number dp=epsilon*abs(pdp);
		
	if(dp==0.0) dp = epsilon;		// avoid numerical error	
	pdp += dp;			// "
	dp = pdp - *p;		// "

	// get value at new point
	number oldp=*p;
	*p=pdp;

	Parameter herep = f.get_parameter("current");

	vect	fudp(f(here));
	
	*p=oldp; // restore parameter
		
	for(integer i=0;i<dims;i++)
	  {
		linpar[i]=(fudp[i]-fu[i])/dp;	
	  }  
  }
  
  template <integer dims, typename nelem, class NT >
  void SSA<dims,nelem, NT>::put_omega(const number& omega)
  {
	for (int i=0;i<dims;i++)
	  {
		A[i][dims+i] = -omega;
		A[dims+i][i] =omega;
	  }
  }
  
}
// end MODEL
#endif 

// CVS Test: hello Guy.

/*********************************************************************
$Id: ssa.h,v 1.4 2002-11-21 09:47:54 mpeeters Exp $
**********************************************************************

$Log: not supported by cvs2svn $
Revision 1.3  2002/11/18 14:10:30  mpeeters
Removed unnecessary ssa.cpp. Added DEBUG in rootscan catch{} routine. Added RC and phase info to the SSA routines.

Revision 1.2  2002/07/30 19:57:19  mpeeters
Merge of 1.0.2 branch into trunk. From now on, the trunk will be a stable (meaning it compiles and has no quirks) branch and all special stuff will be done on separate branches waiting to be merged into the stable one. A tag STABLE will be made which moves with the trunk. Also, the main trunk will be considered to be 1.0.3.

Revision 1.1.2.2  2001/08/30 14:48:27  mpeeters
Testing CVS Branching.

Revision 1.1.2.1  2001/08/30 14:01:22  mpeeters
Fixed stupid warning about extra token (due to #endif SSHA_H).

Revision 1.1  2001/05/22 10:54:55  mpeeters
Moved sources and headers for libModel to model/ subdirectory, in an attempt to rationalize the source tree. This should make things "netter".

Revision 1.1  2001/05/21 11:56:43  mpeeters
Commited lots of small changes (too many to tell) at the previous commit (accidentally). In this commit, we add the small signal analysis stuff. Whoohoo.

*********************************************************************/
